File: trapping-rain-water.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, switching to the 2 pointer method for constant space took some thinking about

/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(heights) {
    if (!heights || !heights.length) return 0;
    let left = 0,
        right = heights.length - 1,
        leftMax = 0,
        rightMax = 0,
        total = 0;
    
    while (left < right) {
        if (heights[left] < heights[right]) {
            if (heights[left] >= leftMax) leftMax = heights[left++];
            else total += leftMax - heights[left++];
        } else {
            if (heights[right] >= rightMax) rightMax = heights[right--];
            else total += rightMax - heights[right--];
        }
    }
    return total;
};

File: TrappingRainWater.cs
using System;
namespace Algorithms
{
    public class TrappingRainWater
    {
        /// Time Complexity : O(N) 
        // Space Complexity :O(1) 
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No 
        public int Trap(int[] height)
        {
            int left = 0;
            int right = height.Length - 1;
            int left_max = 0;
            int right_max = 0;
            int ans = 0;
            while (left < right)
            {
                if (height[left] < height[right])
                {
                    if (height[left] >= left_max)
                    {
                        left_max = height[left];
                    }
                    else
                    {
                        ans += left_max - height[left];
                    }
                    left++;
                }
                else
                {
                    if (height[right] >= right_max)
                    {
                        right_max = height[right];
                    }
                    else
                    {
                        ans += right_max - height[right];
                    }
                    right--;
                }
            }
            return ans;
        }
    }
}
